File: platform/pylib/scout_common/errors/codes.py
"""Error code type hints and utilities for Scout platform errors."""

from typing import NewType


# Code represents a specific error code.
# Codes are string identifiers that uniquely identify error conditions.
#
# Naming convention (recommended but not enforced):
#   - Use UPPER_SNAKE_CASE
#   - Format: ENTITY_CONDITION or OPERATION_FAILURE
#   - Examples: "USER_NOT_FOUND", "DATABASE_CONNECTION_FAILED"
#
# Services should define their own code constants:
#
#     # services/incidents/errors/codes.py
#     from scout_common.errors import Code
#
#     INCIDENT_NOT_FOUND: Code = Code("INCIDENT_NOT_FOUND")
#     INVALID_SEVERITY: Code = Code("INVALID_SEVERITY")
#     GEOCODING_FAILED: Code = Code("GEOCODING_FAILED")
#
Code = NewType("Code", str)


def code(value: str) -> Code:
    """
    Create a Code from a string.
    
    This is a convenience function for creating codes without the NewType constructor.
    
    Args:
        value: The error code string
        
    Returns:
        A typed Code value
        
    Example:
        >>> error_code = code("USER_NOT_FOUND")
        >>> assert error_code == "USER_NOT_FOUND"
    """
    return Code(value)


def is_empty(code_value: Code) -> bool:
    """
    Check if a code is empty.
    
    Args:
        code_value: The code to check
        
    Returns:
        True if the code is an empty string, False otherwise
        
    Example:
        >>> assert is_empty(code(""))
        >>> assert not is_empty(code("USER_NOT_FOUND"))
    """
    return code_value == ""


# Type alias for clarity in type hints
CodeValue = Code

File: platform/pkg/errors/codes.go
// platform/pkg/errors/codes.go
package errors

// Code represents a specific error code.
// Codes are string identifiers that uniquely identify error conditions.
//
// Naming convention (recommended but not enforced):
//   - Use UPPER_SNAKE_CASE
//   - Format: ENTITY_CONDITION or OPERATION_FAILURE
//   - Examples: "USER_NOT_FOUND", "DATABASE_CONNECTION_FAILED"
//
// Services should define their own code constants:
//
//	const (
//	    CodeIncidentNotFound errors.Code = "INCIDENT_NOT_FOUND"
//	    CodeInvalidSeverity  errors.Code = "INVALID_SEVERITY"
//	)
type Code string

// String returns the string representation of the Code.
func (c Code) String() string {
	return string(c)
}

// IsEmpty returns true if the code is empty.
func (c Code) IsEmpty() bool {
	return c == ""
}
